MathUtils have been cleaned and improved

loumalouomega opened the pull request:
Adding new header to the geometries and changing KRATOS_ERROR

loumalouomega committed:
Adding new header to the geometries and changing KRATOS_ERROR

MathUtils have been cleaned and improved too

RiccardoRossi commented on kratos/utilities/math_utils.h:
```diff
@@ -62,180 +27,471 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-    /** Constructor.
-             */
+    /* Constructor */
```
The syntax /** ... */ is for Doxygen comments. It should be preserved!

RiccardoRossi commented on kratos/utilities/math_utils.h:
```diff
@@ -62,180 +27,471 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
     }
 
-    //***********************************************************************
     /**
```
why did u remove InvertMatrix??

RiccardoRossi commented on kratos/utilities/math_utils.h:
```diff
@@ -62,180 +27,471 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+     * @return InvertMatrix: The inverted matrix
+     */
 
-        if(size==1)
```
you can NOT change the API!!

the function InvertMatrix was accepting a matrix of variable size.
The new function you added accepts a matrix of fixed size.

while both functions may be useful you must preserve the original one.

also for the case of fixed-size matrices u should define 3 functions  accepting matrices of size 2*3, 3*3 etc, since this way the if can be avoided

RiccardoRossi commented on kratos/utilities/math_utils.h:
```diff
@@ -475,154 +717,220 @@ class MathUtils
     static inline void  AddMatrix(
         MatrixType& Destination,
         MatrixType& InputMatrix,
         int InitialRow,
```
you could make const the InputMatrix and the integer indices (this one is only a suggestion)

RiccardoRossi commented on kratos/utilities/math_utils.h:
```diff
@@ -777,284 +1080,439 @@ class MathUtils
-      else if (rStrainVector.size()==6)
+        else if (rStrainVector.size()==6)
         {
-	  StrainTensor.resize(3,3, false);
```
here it should be:

if(StrainTenstor.size1() != 3, StrainTensor.size2() != 3)
             StrainTensor.resize(3,3,false);

(clearly this is only an improvement....)

RiccardoRossi commented on kratos/utilities/math_utils.h:
```diff
@@ -1,50 +1,15 @@
-/*
```
My first and most important comment is that this function is used in a lot of places, most of which have no tests.
for this reason if you like to change (and i am positive you do)
you should

STEP1 - create a test written in c++ that tests all of the methods
STEP2 - change it
STEP3 - verify that the tests still run
STEP4 - push the change.

I can not accept a change to this without u going through this procedure... but i hope it is not TOO time consuming to do so.

@roigcarlo could u explain to @loumalouomega how to create and run a test in c++?

RiccardoRossi reviewed (changes requested):
there are a few changes and at least one important requirment (tests)

loumalouomega commented:
The invert matrix is still there, but in other place, not function was removed, and some were added, I can do a test in C++ without problem. Very important, don't look directly to the diff because I moved a lot of things

RiccardoRossi commented:
well...my review was based on diffs so i may be wrong about the removed
functions.

still the main concern about tests is valid.

tests should first pass with the old version, and then pass also with the
new version



<http://www.cimne.com/>

loumalouomega commented:
So, I create a test and I run both in the new and old?,  should I do test for every function or only for the inversion?

loumalouomega commented:
Any example in C++ of tests

RiccardoRossi commented:
for every function... sorry

here you have some examples:

https://github.com/KratosMultiphysics/Kratos/tree/master/kratos/tests/processes

<http://www.cimne.com/>

loumalouomega commented:
I will do, because it is the proper thing to do, but this is not fair, because I just added some functions and the rest remains the same, I just added descriptions, and we didn't have any test previously.

RiccardoRossi commented:
well,
       i can understand your disconfort since it is a lot of work, however
introducing an error in math_utils may break something silently through the
whole kratos,so we need to be extra careful.

I would say that you should test every function you touched, however since
you touched all the file the answer stays more or less the same.

is there a test you see as particularly time consuming? maybe i can help
with that specifically


<http://www.cimne.com/>

loumalouomega commented:
![](https://s-media-cache-ak0.pinimg.com/736x/cc/ab/2d/ccab2dbbb3f5c3c2a9f4a20f8e4857a0.jpg)

loumalouomega committed:
Merge branch 'master' into cleaning-math-utility

maceligueta commented:
I propose to remove some of the functions that calculate the Cross Product, in particular, these ones:
```
    static inline array_1d<double, 3> UnitCrossProduct(const array_1d<double, 3>& vec, const array_1d<double, 3>& Tuple)
    {
        array_1d<double, 3> cross;

        cross[0] =  Tuple[1]*vec[2] - Tuple[2]*vec[1];
        cross[1] =  Tuple[2]*vec[0] - Tuple[0]*vec[2];
        cross[2] =  Tuple[0]*vec[1] - Tuple[1]*vec[0];

        const double length = std::sqrt(inner_prod(cross, cross));
        cross = (1.00/length) * cross;
        return cross;
    }

    static inline array_1d<double, 3> CrossProduct(const array_1d<double, 3>& vec, const array_1d<double, 3>& Tuple)
    {
        array_1d<double, 3> cross;

        cross[0] =  Tuple[1]*vec[2] - Tuple[2]*vec[1];
        cross[1] =  Tuple[2]*vec[0] - Tuple[0]*vec[2];
        cross[2] =  Tuple[0]*vec[1] - Tuple[1]*vec[0];
 
	return cross;
    }
```
Apart from the fact that they return 'b x a' instead of 'a x b', they are slow, because they create a local variable inside the function and probably create a temporary variable when returning the result. If nobody disagrees I can look for all the places where these functions are called and change them in order to call the other versions of the CrossProduct.

loumalouomega commented:
I don't understand how this c++ tests are used or how are they called in python, they are not added to python and I don't understand how they work or even if they are working right now

loumalouomega committed:
Adding test (not working)

RiccardoRossi commented:
+1 for the proposal of MA
<http://www.cimne.com/>

RiccardoRossi commented:
@roigcarlo  (or @pooyan-dadvand) 

we need your help in setting up the c++ test.

also it would be good to have a minimal documentation in the wiki on how to do that

loumalouomega commented:
@maceligueta Agree for your proposals, maybe is a good idea too to create a wishlist of desired functions in the core MathUtils:

- [x] Eigenvalues spectral decomposition (I already added this)
- [ ] Solve second order and third order equations (Cardano formula)
- [ ] QR factorization, or LU for example
- [ ] Tensor operations (I implemented a tensor intersection in the metric utilities of the meshing application, we can move that)

We can add more ideas here

RiccardoRossi commented:
@loumalouomega 
this is just to say that QR decomposition is available in a different header in the core.
@rubenzorrilla any comment on the QR stuff?

also that Matrix inversion by LU is available in ublas.
We could have the InvertMatrix function to fallback to that.

In any case i would add this enhancements as a feature once we close the branck under discussion.
They also would need testing etc etc...

loumalouomega commented:
@RiccardoRossi Agree, but for the tests I need help, doing `greep`s looks like they are implemented but not being used, at least in the nightly compilation with the python script.

RiccardoRossi commented:
ust checking: in order to run the c++ test you must first of all compile them which is done by setting
-DKRATOS_BUILD_TESTING=ON
in the configure script.

having said this, i just had a look to the CMakeFile.txt within the kratos subdir and at line 142-143 the generation of the executable is commented. 
At this point i think we need to ask for help to @roigcarlo  and @pooyan-dadvand

roigcarlo commented:
Hi all, I see some confusion with the tests:

```CMake
if(${KRATOS_BUILD_TESTING} MATCHES ON)
	file(GLOB_RECURSE KRATOS_TESTING_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/tests/*.cpp)
endif(${KRATOS_BUILD_TESTING} MATCHES ON)
```

Tests are compiled here. As you can see everything inside tests directory that its a cpp is compiled.
@RiccardoRossi I think it is safe to remove lines 142-143. Actually there is a LOT of commented code in Kratos that probably should disappear.

To run them ( I don't have the code in hand so I will edit this if it turns to be wrong ) you must;

```Python
from KratosMultiphysics import *

Tester.RunAllTests() #Test all cases
```

RiccardoRossi commented:
Hi Charlie,

i added the line 142-143 some time back since Pooyan wanted to run the tests without the python interface.

anyway, your solution should do the job for us

roigcarlo commented:
@RiccardoRossi Oh, I see, I didn't know, well maybe its good to have something that can run directly from C++. Honestly I don't know.

Anyway, the full interface for the C++ tests is here: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_testing_to_python.cpp

loumalouomega commented:
I cannot compile in using this testing mode, I tried formatting with http://format.krzaq.cc/, but the result is the same, the error that appears is `error: expression cannot be used as a function`

loumalouomega committed:
Trying to solve -DKRATOS_BUILD_TESTING=ON, not solved

loumalouomega commented:
I solved the problem, apparently I needed to add includes that weren't necessaries without testing

loumalouomega committed:
Compìlation in testing solved

loumalouomega committed:
Finally, test added and working

loumalouomega commented:
Tests added and working

rubenzorrilla commented:
@RiccardoRossi @loumalouomega I use the QR decomposition for matrix inversion in one of my convergence accelerators (MVQN) in the FSI app. I agree on moving it to the MathUtils.

RiccardoRossi commented:
@rubenzorrilla i DO NOT propose moving it to math_utils, just suggesting that Vicente can ask you about that feature

RiccardoRossi commented on kratos/tests/math_utils/test_math_utils.cpp:
```diff
@@ -0,0 +1,369 @@
+    namespace Testing 
+    {
+        constexpr double EPSILON = std::numeric_limits<double>::epsilon();
+        constexpr double TOLERANCE = 1e-6;
```
why do you set the tolerance here? i would set it test by test.

Also it should NOT be defined for the whole "Testing" namespace

RiccardoRossi commented on kratos/tests/math_utils/test_math_utils.cpp:
```diff
@@ -0,0 +1,369 @@
+
+            KRATOS_CHECK_NEAR(det, 1.0, TOLERANCE);
+            
+            boost::numeric::ublas::bounded_matrix<double, 2, 2> mat22 = ZeroMatrix(2, 2);
```
well...it would be good something a little less trivial...

like 1,2
      3,4

RiccardoRossi commented on kratos/tests/math_utils/test_math_utils.cpp:
```diff
@@ -0,0 +1,369 @@
+            boost::numeric::ublas::bounded_matrix<double, 1, 1> mat11 = ZeroMatrix(1, 1);
+            mat11(0,0) = 1.0;
+            
+            double det = MathUtils<double>::DetMat<1>(mat11);
```
Is this "DetMat" function new?
if so, please give a more expressive name

RiccardoRossi commented on kratos/tests/math_utils/test_math_utils.cpp:
```diff
@@ -0,0 +1,369 @@
+            mat44(3,3) = 3.0;
+            
+            boost::numeric::ublas::bounded_matrix<double, 4, 4> inv44 = MathUtils<double>::InvertMatrix<4>(mat44, det);
+
```
i would say that here tolerance should be MUCH stricter than 1e-6

(i mean in all of this test)

RiccardoRossi commented on kratos/tests/math_utils/test_math_utils.cpp:
```diff
@@ -0,0 +1,369 @@
+        {
+            int sign = MathUtils<double>::Sign(-1.0);
+            
+            KRATOS_CHECK_EQUAL(sign, -1.0);
```
you should check equality to an integer no? i mean
KRATOS_CHECK_EQUAL(sign, -1);

RiccardoRossi commented on kratos/tests/math_utils/test_math_utils.cpp:
```diff
@@ -0,0 +1,369 @@
+        
+    } // namespace Testing
+}  // namespace Kratos.
+
```
what about the cross_product functions?

the norm?

i think that you are still missing some funcitons

RiccardoRossi replied:
warnings to be removed

/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp: In member function ‘virtual void Kratos::Testing::TestMathUtilsDetMatTest::TestFunction()’:
/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:93:65: warning: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[4ul]’ is used uninitialized in this function [-Wuninitialized]
             boost::numeric::ublas::bounded_matrix<double, 1, 1> mat11 = ZeroMat
                                                                 ^
In file included from /home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:23:0:
/home/rrossi/kratos_git/kratos/utilities/math_utils.h:203:13: warning: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[2ul]’ is used uninitialized in this function [-Wuninitialized]
             InputMatrixDet = (InputMatrix(0, 0) * InputMatrix(1, 1) * InputMatr
             ^
/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:93:65: note: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[2ul]’ was declared here
             boost::numeric::ublas::bounded_matrix<double, 1, 1> mat11 = ZeroMat
                                                                 ^
In file included from /home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:23:0:
/home/rrossi/kratos_git/kratos/utilities/math_utils.h:203:13: warning: ‘*((void*)& mat11 +32)’ is used uninitialized in this function [-Wuninitialized]
             InputMatrixDet = (InputMatrix(0, 0) * InputMatrix(1, 1) * InputMatr
             ^
/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:93:65: note: ‘*((void*)& mat11 +32)’ was declared here
             boost::numeric::ublas::bounded_matrix<double, 1, 1> mat11 = ZeroMat
                                                                 ^
In file included from /home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:23:0:
/home/rrossi/kratos_git/kratos/utilities/math_utils.h:203:13: warning: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[5ul]’ is used uninitialized in this function [-Wuninitialized]
             InputMatrixDet = (InputMatrix(0, 0) * InputMatrix(1, 1) * InputMatr
             ^
/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:93:65: note: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[5ul]’ was declared here
             boost::numeric::ublas::bounded_matrix<double, 1, 1> mat11 = ZeroMat
                                                                 ^
In file included from /home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:23:0:
/home/rrossi/kratos_git/kratos/utilities/math_utils.h:203:13: warning: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[3ul]’ is used uninitialized in this function [-Wuninitialized]
             InputMatrixDet = (InputMatrix(0, 0) * InputMatrix(1, 1) * InputMatr
             ^
/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:93:65: note: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[3ul]’ was declared here
             boost::numeric::ublas::bounded_matrix<double, 1, 1> mat11 = ZeroMat
                                                                 ^
/home/rrossi/kratos_git/kratos/tests/math_utils/test_math_utils.cpp:93:65: warning: ‘*((void*)(& mat11)+16).boost::numeric::ublas::bounded_array<double, 1ul, std::allocator<double> >::data_[6ul]’ is used uninitialized in this function [-Wuninitialized]

loumalouomega replied:
The norms and cross products are there

RiccardoRossi commented:
failure in tests:

    TestMathUtilsDetMatTest Failed with message: 
        Error: Check failed becuase det = 0 is not near to 1.0 = 1 within the tolerance 1e-06
in kratos/tests/math_utils/test_math_utils.cpp:98:virtual void Testing::TestMathUtilsDetMatTest::TestFunction()

    TestMathUtilsInvMatTest Failed with message: 
        Error:  Determinant of the matrix is zero or almost zero!!!, InputMatrixDet = -5.99724e-61

in kratos/utilities/math_utils.h:277:static boost::numeric::ublas::bounded_matrix<TDataType, TDim, TDim> MathUtils<TDataType>::InvertMatrix(const boost::numeric::ublas::bounded_matrix<TDataType, TDim, TDim>&, TDataType&, TDataType) [with unsigned int TDim = 1u; TDataType = double]

    TestModelPartIOSubModelPartsDivision Failed with message: 
        Error: The nodal solution step container deos not have this variable: VELOCITY

loumalouomega commented:
I am not responsible of TestModelPartIOSubModelPartsDivision

loumalouomega commented:
@RiccardoRossi  and @pooyan-dadvand I created the wiki for C++ tests, https://github.com/KratosMultiphysics/Kratos/wiki/How-to-create-unitary-tests#c, please, when something new is created the proper thing is to write the documentation before asking to use it. I was the Saturday night trying to figure out and at the end was a very simple action

loumalouomega committed:
Adding more tests and correcting 4x4 matrix inversion

RiccardoRossi commented:
ok, i can now accept it

thx

RiccardoRossi reviewed (approved):
ok, approved

loumalouomega committed:
Minor

pooyan-dadvand commented:
@loumalouomega thank you very much for the wiki page! 🥇
